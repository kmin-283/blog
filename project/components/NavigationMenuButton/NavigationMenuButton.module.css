.navMenuButton {
    display: none;
}

@media (max-width: 799px) {
    .navMenuButton {
        display: block;
        position: fixed;
        bottom: 0;
        width: 100%;
        padding: 1rem 0;
        border-right: 2px solid var(--main-theme-color);
        border-left: 2px solid var(--main-theme-color);
        background-color: var(--admin-background-color);
    }

    .navMenuButton div {
        display: flex;
        justify-content: center;
        width: 100%;
        max-height: 20vh;
        overflow: auto;
    }

    .button {
        width: 100%;
        border: none;
        cursor: pointer;
        font-size: 1.5rem;
        font-weight: 700;
    }

    .button p {
        margin: 2px;
    }
}
