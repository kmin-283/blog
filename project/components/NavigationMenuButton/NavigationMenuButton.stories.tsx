import React from 'react';
import { Story, Meta } from '@storybook/react';
import NavigationMenuButton, {NavigationMenuButtonProps} from '@/components/NavigationMenuButton/NavigationMenuButton';

export default {
  title: 'Components/NavigationMenuButton',
  component: NavigationMenuButton,
} as Meta;

const Template: Story<NavigationMenuButtonProps> = ({internalLinks}) => {
  
  return (
    <div style={{position:'relative', width: 500, height: 300}}>
      <NavigationMenuButton internalLinks={internalLinks}>
        버튼을 누르면 나옵니다
      </NavigationMenuButton>
    </div>);
};

export const BasicNavigationMenuButton = Template.bind({});
BasicNavigationMenuButton.args = {
  internalLinks: "[" +
    "{\"headingLevel\":2,\"value\":\"1 헤딩\",\"child\":[]}," +
    "{\"headingLevel\":2,\"value\":\"2 헤딩\",\"child\":[" +
    "{\"headingLevel\":3,\"value\":\"2-1 헤딩\",\"child\":[" +
    "{\"headingLevel\":4,\"value\":\"2-1-1 헤딩\",\"child\":[]}," +
    "{\"headingLevel\":4,\"value\":\"2-1-2 헤딩\",\"child\":[" +
    "{\"headingLevel\":5,\"value\":\"2-1-2-1 헤딩\",\"child\":[]}]}," +
    "{\"headingLevel\":4,\"value\":\"2-1-3 헤딩\",\"child\":[]}]}," +
    "{\"headingLevel\":3,\"value\":\"2-2 헤딩\",\"child\":[]}]}]\n",
};