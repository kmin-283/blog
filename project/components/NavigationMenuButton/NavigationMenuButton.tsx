import React, {FC, useState} from 'react';
import HierarchicalList from '@/components/HierarchicalList/HierarchicalList';
import styles from './NavigationMenuButton.module.css';

export interface NavigationMenuButtonProps {
  internalLinks: string;
}

const NavigationMenuButton: FC<NavigationMenuButtonProps> = ({internalLinks, children}) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const openMenu = () => {
    setIsOpen(prevState => !prevState);
  };
  return (
    <div className={styles.navMenuButton}>
      <button className={styles.button} id='menubutton' aria-haspopup='true' aria-expanded={isOpen} onClick={openMenu}>
        <p>{children}</p>
      </button>
      {isOpen &&
      (<div>
        <HierarchicalList role='menu' labelledBy='menubutton' internalLinks={internalLinks}/>
      </div>)}
    </div>
  );
};

export default NavigationMenuButton;
