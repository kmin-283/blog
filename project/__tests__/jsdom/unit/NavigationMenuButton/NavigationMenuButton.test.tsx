import React from "react";
import NavigationMenuButton from "@/components/NavigationMenuButton/NavigationMenuButton";
import {
  render,
  cleanup,
  screen,
  fireEvent,
  RenderResult,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

describe("NavigationMenuButton 컴포넌트", () => {
  const internalLinks =
    "[" +
    '{"headingLevel":2,"value":"1 헤딩","child":[]},' +
    '{"headingLevel":2,"value":"2 헤딩","child":[' +
    '{"headingLevel":3,"value":"2-1 헤딩","child":[' +
    '{"headingLevel":4,"value":"2-1-1 헤딩","child":[]},' +
    '{"headingLevel":4,"value":"2-1-2 헤딩","child":[' +
    '{"headingLevel":5,"value":"2-1-2-1 헤딩","child":[]}]},' +
    '{"headingLevel":4,"value":"2-1-3 헤딩","child":[]}]},' +
    '{"headingLevel":3,"value":"2-2 헤딩","child":[]}]}]\n';
  let componenet: RenderResult;

  afterAll(cleanup);
  beforeEach(() => {
    componenet = render(<NavigationMenuButton internalLinks={internalLinks} />);
  });

  test("렌더링", () => {
    const navigationMenuButton = renderer
      .create(<NavigationMenuButton internalLinks={internalLinks} />)
      .toJSON();
    expect(navigationMenuButton).toMatchSnapshot();
  });

  test("클릭 이전엔 항목이 보이지 않는다", () => {
    expect(screen.queryByRole("ol")).not.toBeInTheDocument();
  });

  test("클릭 이후 항목이 보여야 한다", () => {
    const { getByRole } = componenet;
    const btn = getByRole("button");

    fireEvent.click(btn);

    const menu = getByRole("menu");

    expect(menu).toBeInTheDocument();
    expect(menu.children).toHaveLength(2);
  });
});
