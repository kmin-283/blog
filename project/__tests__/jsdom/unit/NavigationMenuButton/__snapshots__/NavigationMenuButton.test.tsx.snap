// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationMenuButton 컴포넌트 렌더링 1`] = `
<div
  className="navMenuButton"
>
  <button
    aria-expanded={false}
    aria-haspopup="true"
    className="button"
    id="menubutton"
    onClick={[Function]}
  >
    <p />
  </button>
</div>
`;
